ReSharper.FSharp/src/FSharp.Psi.Features/src/Daemon/Stages/InferredTypeHintStage.fs

```diff
@@ -0,0 +1,63 @@
+﻿namespace JetBrains.ReSharper.Plugins.FSharp.Psi.Features.Daemon.Stages
+
+open JetBrains.Application.Settings
+open JetBrains.ProjectModel
+open JetBrains.ReSharper.Daemon.Stages
+open JetBrains.ReSharper.Plugins.FSharp
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Features.Daemon.Highlightings
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Impl
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Impl.Tree
+open JetBrains.ReSharper.Plugins.FSharp.Settings
+open JetBrains.ReSharper.Psi
+open JetBrains.ReSharper.Feature.Services.Daemon
+open JetBrains.ReSharper.Plugins.FSharp.Daemon.Cs.Stages
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Tree
+open FSharp.Compiler.SourceCodeServices
+
+type InferredTypeHintHighlightingProcess(logger: ILogger, fsFile, settings: IContextBoundSettingsStore, daemonProcess: IDaemonProcess) =
+    inherit FSharpDaemonStageProcessBase(fsFile, daemonProcess)
+
+    override x.Execute(committer) =
+        let consumer = new FilteringHighlightingConsumer(daemonProcess.SourceFile, fsFile, settings)
+        fsFile.ProcessThisAndDescendants(Processor(x, consumer))
+        committer.Invoke(DaemonStageResult(consumer.Highlightings))
+
+    override x.VisitLocalReferencePat(localRefPat, consumer) =
+        match localRefPat.IgnoreParentParens().Parent with
+        | :? ITypedPat -> ()
+        | :? ILocalBinding as localBinding when localBinding.ReturnTypeInfo <> null -> ()
```

Comment:
It would probably be better to use generated navigators here, since using them would allow change the tree structure in future, as it's still shaping up.

Maybe something like that:
```fsharp
let pat = localRefPat.IgnoreParentParens()
if isNotNull (TypedPatNavigator.GetByPattern(pat)) then () else

let binding = BindingNavigator.GetByHeadPattern(pat)
if isNotNull binding && isNotNull binding.ReturnTypeInfo then () else
```